rxnetty-common/src/main/java/io/reactivex/netty/util/CollectBytes.java

```
@@ -0,0 +1,153 @@
+                return upstream
+                        .compose(new TransformerLimitByteBufTotalSize(maxBytes))
+                        .compose(INSTANCE)
+                        .flatMap(new Func1<ByteBuf, Observable<ByteBuf>>() {
+                            @Override
+                            public Observable<ByteBuf> call(ByteBuf buf) {
+                                return buf.readableBytes() > maxBytes ?
+                                        Observable.<ByteBuf>error(OnErrorThrowable.addValueAsLastCause(new TooMuchDataException(), buf)) :
+                                        Observable.just(buf);
+                            }
+                        });
+            }
+        };
+    }
+
+    private CollectBytes() {
+        // do nothing
+    }
+
+    @Override
+    public Observable<ByteBuf> call(Observable<ByteBuf> upstream) {
+        return upstream.collect(
+                new Func0<CompositeByteBuf>() {
+                    @Override
+                    public CompositeByteBuf call() {
+                        return Unpooled.compositeBuffer();
+                    }
+                },
+                new Action2<CompositeByteBuf, ByteBuf>() {
+                    @Override
+                    public void call(CompositeByteBuf collector, ByteBuf buf) {
+                        int i = buf.readableBytes();
+                        collector.addComponent(buf);
+                        collector.writerIndex(collector.writerIndex() + i);
+                    }
+                }
+            )
+            .cast(ByteBuf.class);
+    }
+
+    public static class TransformerLimitByteBufTotalSize implements Observable.Transformer<ByteBuf, ByteBuf> {
+
+        private final int maxBytes;
+
+        public TransformerLimitByteBufTotalSize(int maxBytes) {
+            this.maxBytes = maxBytes;
+        }
+
+        @Override
+        public Observable<ByteBuf> call(Observable<ByteBuf> upstream) {
+            return upstream.scan(
```

@jamesgorman2 what's your take on this comment i.e. just do `collect()` instead of `scan+skip+takeUntil` and throw when the size limit is breached.